### DISCUSSION TITLE
Chemical composition mixtures and not equilibrium transformations

### DISCUSSION THREAD

**Discussion by gplanu** [pycalphad/pycalphad#437] [Q&A]
Dears,
after a very fruitful discussion also with Mr. Richard Otis, I am facing to these two problems next explained:
1. Evaluation of complex multi component chemical composition: how could be the fastest way to extract for a defined composition the resulted Aem, A1 equilibrium curves using _calculate_ and _equilibrium_ pycalphad methods? Is it possible to obtain the resulted binary diagram maintaining constant i.e. Mn, V, Nb and varying only Fe and C (with vacancies of course)?
2. Regarding martensite and bainite transformation working with the same complex mixtures I have convergence problems adding components. I am working using an second order explicit Euler algorithm (I attach a pic below). Red points are using a simplified algorithm in Matlab, blue ones are from pycalphad. 
![example](https://user-images.githubusercontent.com/117755633/200613270-263e33a4-6fb7-44ad-918e-e1196d5d163c.png)
![example1](https://user-images.githubusercontent.com/117755633/200613593-8bd07f36-56e3-4f75-b4ba-7e6449fc1da0.png)
![example2](https://user-images.githubusercontent.com/117755633/200613594-60e5bc31-c88f-4ee1-8ccc-28a4ad5b2d7e.png)
![example3](https://user-images.githubusercontent.com/117755633/200613597-3560c72c-d40a-4400-b594-47981b6acab6.png)

Looking forward for your suggestions.
Thanks a lot in advance for your help and attention.
Best regards!

**Comment by richardotis** [comment]
1. At the moment, a binary iterative search is probably the quickest way to find liquidus or solidus temperatures. For binary phase diagrams, it's possible to programmatically access the phase boundaries directly, but a binary search is going to be the fastest general way for now. For isopleth diagrams, pycalphad does not offer any support yet. We have something in progress for it, but it will probably not be ready before the new year.
2. We have some bug fixes ready to be merged which may address this issue.

**Reply by gplanu** [reply]
Thanks again Richard for your kind help and attention to this topic. Actually regarding the two topic:
1. I report my code, is not so slow using directly this procedure as for example: 
conditions = {v.P: 101325,
              v.T: (300, 2000, 50),
              v.X('C'):(0,0.25,0.01),
              v.X('MN'):round(chemComp['X_MN'],5),
              v.X('V'):round(chemComp['X_V'],5),
        } 
G_n = equilibrium(dbf,comps,["FCC_A1","BCC_A2"], conditions)

In any case how do you suggest to extract G_n fields such us Phase, X_C, Temperatures without work inside equilibrium.py and with ZPF_boundary_sets?

2. When do you have planned to release the merged version?

Thanks a lot for your effort and help!
Best regards.

**Reply by gplanu** [reply]
Dear Richard, 
inside the object all_phase_data, evaluating the calculate method, which one is the right composition? The X or Y? 
Moreover, the GM values are in rows the energies according to different composition mixtures for each temperature right?
![all_phase_data](https://user-images.githubusercontent.com/117755633/201142877-a2b95ce9-bbb3-4f67-9968-8353f757d5fb.PNG)

**Reply by richardotis** [reply]
The documentation has an explanation of what all the arrays mean: https://pycalphad.org/docs/latest/examples/UsingCalculationResults.html

**Reply by gplanu** [reply]
Thanks Richard, but from your point of view which are the best to keep? I mean if you see the screen above the composition X does not uses always all the 4 components, the site fraction yes. So I do not understand which is the right to consider for the evaluation of Ae3 and Aem of complex mixtures. Thanks a lot for your patience and help.

**Reply by richardotis** [reply]
The calculate function generates a random sample of the composition space. `X` is the mole fraction of the configuration, and `Y` are the sublattice site fractions (only relevant for multi-sublattice phases). If you want the minimum energy configuration for a given phase composition, you should use the `equilibrium` function and step in the composition space, like
```
eq = equilibrium(dbf, comps, [phase_name], {v.T: 600, v.P: 1e5, v.N: 1, v.X('A'): (0,1,0.01)})
```

**Reply by gplanu** [reply]
Thanks Richard,
in any case debugging the equilibrium method it is called also the calculate one...
For the suggested code, yes I could proceed in this way and cycle for each temperature and carbon fraction (maintaining constants the other mixture elements) in order to find the minimum energy for each phase and rebuild the simplified phase diagram. 
What do you think?

**Reply by richardotis** [reply]
I think I understand your question a little better now. PyCalphad doesn't support isopleth diagram calculation yet. The brute force way to do what you want (I think you already know this) is to perform an equilibrium calculation at each composition and temperature of interest, and then use the `eq.Phase` and `eq.X` arrays to plot the phase boundary manually. For convergence issues, you might also try adjusting the `pdens` keyword argument; in the next version of pycalphad this should be unnecessary.

**Reply by gplanu** [reply]
Yes, you perfectly understand my point. I agree with the strategy you have proposed. Could you only give me any support with _pdens_ argument? How can I adjust it? In your trials which is the best trade-off of elements to use for a mixture?
Thanks a lot, it is a pleasure to work together!

**Reply by richardotis** [reply]
The default pdens value is 60. It has sort of a complicated meaning, but increasing it will increase the number of starting points during minimization. Try `equilibrium(..., grid_opts={'pdens': 120})` to start.

**Reply by gplanu** [reply]
Dear Richard @richardotis, I update you regarding my situation, hoping to help you and to have new suggestions.
1. According to a defined composition attached below, I have extracted the equilibrium for: FCC_A1, BCC_A2, CEMENTITE 
![composition](https://user-images.githubusercontent.com/117755633/202262819-ba2d47bd-5710-430e-a4d5-299fb567c6cb.PNG)
Of course for the first 2 phases we have a very large quadratic curve (fitted in its points) and for cementite a line according to each temperature for x_carbon @ 0.25.
Here the example of the phase values extraction:
![image](https://user-images.githubusercontent.com/117755633/202265770-541f8180-2bfa-49e1-8747-ac4bc7b609b2.png)

2. So I have calculated the tangent line for the two quadratic curves in order to find the X_C for the tie line (so for each temperature).
According to that I found two facts:
               a. NaN values sometimes are calculated (attachments below). Do you know the reason? Is it physical meaning or numerical 
                   meaning?
![bcc](https://user-images.githubusercontent.com/117755633/202264776-4ec5ba19-904a-4def-b314-4ebcaaa0b11b.PNG)
![fcc](https://user-images.githubusercontent.com/117755633/202264793-cbdd1376-2a82-403e-94d4-b46acefe023d.PNG)

               b. Values are not calculated for small amounts of Carbon (plot below), comparing to simplified matlab extraction (same .tdb, 
                   considered only the carbon contributes) has a very different profile. (blue from pycalphad, red matlab)
![compare](https://user-images.githubusercontent.com/117755633/202265203-b20c2803-8c04-46b1-9c54-52829dce2307.PNG)

What do you think about this approach to finally extract Ae3 and Aem for complex mixture? Do you have any suggestion regarding the NaN outputs and regarding the differences between pycalphad and matlab? Am I right following this approach? 

Thanks in advance for your help.
Best Regards

**Reply by richardotis** [reply]
One reason for the discrepancy could be due to a miscibility gap. Have you confirmed that each of your calculations is only forming a single phase at a time?

**Reply by gplanu** [reply]
The excel screens above show that there is the phase formation. Moreover, I tried to launch an equilibrium calculation of the two phases and the result is in the screen below:
![image](https://user-images.githubusercontent.com/117755633/202383720-e806c39a-8514-4b11-bb71-0704cb396a00.png)
So is not easy to understand which phase has to be considered... that is the reason why I came back to the theory to fit the quadratic equation..

If you see the composition is really a steel with low amount of manganese and vanadium, so I expect that the calculation performs without any issue.

**Reply by gplanu** [reply]
Dear Richard, regarding our conversation attached below in pycalphad/pycalphad Gitter: which is the main difference between your suggestion and the tentative that I show you above considering equilibrium method giving as input a single phase? 
Could you explain me please that from calphad and numerical point of view? Thanks a lot!

:point_up: [Edit](https://gitter.im/pycalphad/pycalphad?at=631287eab16e8236e35b414a): @gplalab_gitlab: I have some new features in progress which should address this concern. For T0, I have some code using my development branch which uses the entropy differences between the phases to find it efficiently for a given composition. In the next version you'll also be able to access the CompositionSet objects and control the energy minimizer directly, without automated miscibility gap detection interfering.

The current concept (subject to change) looks like:

# Get a starting point from a previous calculation, called 'wks' (not shown)
def find_first_compset(phase_name):
    for index, compsets in wks.enumerate_composition_sets():
        for compset in compsets:
            if compset.phase_record.phase_name == phase_name:
                return compset
    return None
bcc_compset = find_first_compset('BCC_A2')
hcp_compset = find_first_compset('HCP_A3')
# [ some code omitted for brevity ]
bcc_result = Solver().solve([bcc_compset], initial_conditions)
hcp_result = Solver().solve([hcp_compset], initial_conditions)
The solver won't add any phases when used this way, so you can manipulate single-phase properties.

**Reply by gplanu** [reply]
Hello @richardotis, I have another question for you. I am solving an equilibrium for a steel mixture with 10 components using equilibrium method. I am trying to solve equilibrium only for the BCC_A2 phase with a range of 300-2000 K (temperature step 50K) and carbon range 0-0.25 molar fraction (carbon step 0.01).
During phase elaboration I have this screen which reports to solver.py (attached). Any suggestion?
![image](https://user-images.githubusercontent.com/117755633/203847511-508060e8-8d32-4890-b70b-a4f769f8a292.png)

Thanks a lot for your help.
Giulio

**Reply by richardotis** [reply]
That looks like a bug. If you can put together a reproducible example, you can open a new issue so we can address it.

**Reply by gplanu** [reply]
Ok, I will share in a different issue. In any case, I think is a bug because working with single temperatures it solve (of course with NaN case) but few points could be extracted.
